File: client/src/main/scala/stasis/client/tracking/state/RecoveryState.scala
package stasis.client.tracking.state

import stasis.client.model.TargetEntity.Destination
import stasis.client.model.{proto, EntityMetadata, TargetEntity}
import stasis.shared.ops.Operation

import java.nio.file.{Path, Paths}
import java.time.Instant
import scala.util.{Failure, Success, Try}

final case class RecoveryState(
  operation: Operation.Id,
  started: Instant,
  entities: RecoveryState.Entities,
  failures: Seq[String],
  completed: Option[Instant]
) extends OperationState {
  override val `type`: Operation.Type = Operation.Type.Recovery

  override val isCompleted: Boolean = completed.isDefined

  def entityExamined(entity: Path): RecoveryState =
    copy(entities = entities.copy(examined = entities.examined + entity))

  def entityCollected(entity: TargetEntity): RecoveryState =
    copy(entities = entities.copy(collected = entities.collected + (entity.path -> entity)))

  def entityProcessingStarted(entity: Path, expectedParts: Int): RecoveryState =
    copy(
      entities = entities.copy(
        pending =
          entities.pending + (entity -> RecoveryState.PendingTargetEntity(expectedParts = expectedParts, processedParts = 0))
      )
    )

  def entityPartProcessed(entity: Path): RecoveryState =
    copy(entities = entities.copy(pending = entities.pending + (entity -> entities.pending(entity).inc())))

  def entityProcessed(entity: Path): RecoveryState = {
    val processed = entities.pending.get(entity) match {
      case Some(pending) =>
        RecoveryState.ProcessedTargetEntity(
          expectedParts = pending.expectedParts,
          processedParts = pending.processedParts
        )

      case None =>
        RecoveryState.ProcessedTargetEntity(
          expectedParts = 0,
          processedParts = 0
        )
    }

    copy(
      entities = entities.copy(
        pending = entities.pending - entity,
        processed = entities.processed + (entity -> processed)
      )
    )
  }

  def entityMetadataApplied(entity: Path): RecoveryState =
    copy(entities = entities.copy(metadataApplied = entities.metadataApplied + entity))

  def entityFailed(entity: Path, reason: Throwable): RecoveryState =
    copy(entities =
      entities.copy(failed = entities.failed + (entity -> s"${reason.getClass.getSimpleName} - ${reason.getMessage}"))
    )

  def failureEncountered(failure: Throwable): RecoveryState =
    copy(failures = failures :+ s"${failure.getClass.getSimpleName} - ${failure.getMessage}")

  def recoveryCompleted(): RecoveryState =
    copy(completed = Some(Instant.now()))

  override def asProgress: Operation.Progress = Operation.Progress(
    started = started,
    total = entities.examined.size,
    processed = entities.processed.size,
    failures = entities.failed.size + failures.size,
    completed = completed
  )
}

object RecoveryState {
  def start(operation: Operation.Id): RecoveryState =
    RecoveryState(
      operation = operation,
      started = Instant.now(),
      entities = Entities.empty,
      failures = Seq.empty,
      completed = None
    )

  final case class Entities(
    examined: Set[Path],
    collected: Map[Path, TargetEntity],
    pending: Map[Path, PendingTargetEntity],
    processed: Map[Path, ProcessedTargetEntity],
    metadataApplied: Set[Path],
    failed: Map[Path, String]
  )

  object Entities {
    def empty: Entities = Entities(
      examined = Set.empty,
      collected = Map.empty,
      pending = Map.empty,
      processed = Map.empty,
      metadataApplied = Set.empty,
      failed = Map.empty
    )
  }

  final case class PendingTargetEntity(
    expectedParts: Int,
    processedParts: Int
  ) {
    def inc(): PendingTargetEntity =
      copy(processedParts = processedParts + 1)
  }

  final case class ProcessedTargetEntity(
    expectedParts: Int,
    processedParts: Int
  )

  def toProto(state: RecoveryState): proto.state.RecoveryState =
    proto.state.RecoveryState(
      started = state.started.toEpochMilli,
      entities = Some(
        proto.state.RecoveryEntities(
          examined = state.entities.examined.map(_.toAbsolutePath.toString).toSeq,
          collected = state.entities.collected.map { case (k, v) =>
            k.toAbsolutePath.toString -> toProtoTargetEntity(v)
          },
          pending = state.entities.pending.map { case (k, v) =>
            k.toAbsolutePath.toString -> toProtoPendingTargetEntity(v)
          },
          processed = state.entities.processed.map { case (k, v) =>
            k.toAbsolutePath.toString -> toProtoProcessedTargetEntity(v)
          },
          metadataApplied = state.entities.metadataApplied.map(_.toAbsolutePath.toString).toSeq,
          failed = state.entities.failed.map { case (k, v) => k.toAbsolutePath.toString -> v }
        )
      ),
      failures = state.failures,
      completed = state.completed.map(_.toEpochMilli)
    )

  def fromProto(operation: Operation.Id, state: proto.state.RecoveryState): Try[RecoveryState] =
    state.entities match {
      case Some(entities) =>
        Try {
          RecoveryState(
            operation = operation,
            started = Instant.ofEpochMilli(state.started),
            entities = RecoveryState.Entities(
              examined = entities.examined.map(Paths.get(_)).toSet,
              collected = entities.collected.map { case (k, v) => Paths.get(k) -> fromProtoTargetEntity(v) },
              pending = entities.pending.map { case (k, v) => Paths.get(k) -> fromProtoPendingTargetEntity(v) },
              processed = entities.processed.map { case (k, v) => Paths.get(k) -> fromProtoProcessedTargetEntity(v) },
              metadataApplied = entities.metadataApplied.map(Paths.get(_)).toSet,
              failed = entities.failed.map { case (k, v) => Paths.get(k) -> v }
            ),
            failures = state.failures,
            completed = state.completed.map(Instant.ofEpochMilli)
          )
        }

      case None =>
        Failure(new IllegalArgumentException("Expected entities in recovery state but none were found"))
    }

  @SuppressWarnings(Array("org.wartremover.warts.Throw"))
  private def fromProtoTargetEntity(entity: proto.state.TargetEntity): TargetEntity =
    TargetEntity(
      path = Paths.get(entity.path),
      destination = entity.destination match {
        case directory: proto.state.TargetEntityDestinationDirectory =>
          TargetEntity.Destination.Directory(
            path = Paths.get(directory.path),
            keepDefaultStructure = directory.keepDefaultStructure
          )

        case _ =>
          TargetEntity.Destination.Default
      },
      existingMetadata = entity.existingMetadata.flatMap(metadata => EntityMetadata.fromProto(metadata).toOption) match {
        case Some(metadata) => metadata
        case None => throw new IllegalArgumentException("Expected existing metadata in recovery state but none was found")
      },
      currentMetadata = entity.currentMetadata.map { metadata =>
        EntityMetadata.fromProto(metadata) match {
          case Success(metadata) => metadata
          case Failure(e)        => throw e
        }
      }
    )

  private def toProtoTargetEntity(entity: TargetEntity): proto.state.TargetEntity =
    proto.state.TargetEntity(
      path = entity.path.toAbsolutePath.toString,
      destination = entity.destination match {
        case directory: Destination.Directory =>
          proto.state.TargetEntityDestinationDirectory(
            path = directory.path.toAbsolutePath.toString,
            keepDefaultStructure = directory.keepDefaultStructure
          )

        case Destination.Default =>
          proto.state.TargetEntityDestinationDefault()
      },
      existingMetadata = Some(EntityMetadata.toProto(entity.existingMetadata)),
      currentMetadata = entity.currentMetadata.map(EntityMetadata.toProto)
    )

  private def fromProtoPendingTargetEntity(entity: proto.state.PendingTargetEntity): PendingTargetEntity =
    PendingTargetEntity(
      expectedParts = entity.expectedParts,
      processedParts = entity.processedParts
    )

  private def toProtoPendingTargetEntity(entity: PendingTargetEntity): proto.state.PendingTargetEntity =
    proto.state.PendingTargetEntity(
      expectedParts = entity.expectedParts,
      processedParts = entity.processedParts
    )

  private def fromProtoProcessedTargetEntity(entity: proto.state.ProcessedTargetEntity): ProcessedTargetEntity =
    ProcessedTargetEntity(
      expectedParts = entity.expectedParts,
      processedParts = entity.processedParts
    )

  private def toProtoProcessedTargetEntity(entity: ProcessedTargetEntity): proto.state.ProcessedTargetEntity =
    proto.state.ProcessedTargetEntity(
      expectedParts = entity.expectedParts,
      processedParts = entity.processedParts
    )
}

File: client-android/lib/src/main/kotlin/stasis/client_android/lib/tracking/state/RecoveryState.kt
package stasis.client_android.lib.tracking.state

import stasis.client_android.lib.model.EntityMetadata.Companion.toModel
import stasis.client_android.lib.model.EntityMetadata.Companion.toProto
import stasis.client_android.lib.model.EntityRef
import stasis.client_android.lib.model.TargetEntity
import stasis.client_android.lib.ops.Operation
import stasis.client_android.lib.ops.OperationId
import stasis.client_android.lib.utils.Try
import java.nio.file.Paths
import java.time.Instant

data class RecoveryState(
    val operation: OperationId,
    override val started: Instant,
    val entities: Entities,
    val failures: List<String>,
    override val completed: Instant?
) : OperationState {
    override val type: Operation.Type
        get() = Operation.Type.Recovery

    fun entityExamined(entity: EntityRef): RecoveryState =
        copy(entities = entities.copy(examined = entities.examined.plus(element = entity)))

    fun entityCollected(entity: TargetEntity): RecoveryState =
        copy(entities = entities.copy(collected = entities.collected + (entity.ref to entity)))

    fun entityProcessingStarted(entity: EntityRef, expectedParts: Int): RecoveryState =
        copy(
            entities = entities.copy(
                pending =
                entities.pending + (entity to PendingTargetEntity(
                    expectedParts = expectedParts,
                    processedParts = 0
                ))
            )
        )

    fun entityPartProcessed(entity: EntityRef): RecoveryState =
        copy(entities = entities.copy(pending = entities.pending + (entity to entities.pending[entity]!!.inc())))

    fun entityProcessed(entity: EntityRef): RecoveryState {
        val processed = when (val pending = entities.pending[entity]) {
            null -> ProcessedTargetEntity(
                expectedParts = 0,
                processedParts = 0
            )

            else -> ProcessedTargetEntity(
                expectedParts = pending.expectedParts,
                processedParts = pending.processedParts
            )
        }

        return copy(
            entities = entities.copy(
                pending = entities.pending.minus(key = entity),
                processed = entities.processed + (entity to processed)
            )
        )
    }

    fun entityMetadataApplied(entity: EntityRef): RecoveryState =
        copy(entities = entities.copy(metadataApplied = entities.metadataApplied.plus(element = entity)))

    fun entityFailed(entity: EntityRef, reason: Throwable): RecoveryState =
        copy(
            entities = entities.copy(failed = entities.failed + (entity to "${reason.javaClass.simpleName} - ${reason.message}"))
        )

    fun failureEncountered(failure: Throwable): RecoveryState =
        copy(failures = failures + "${failure.javaClass.simpleName} - ${failure.message}")

    fun recoveryCompleted(): RecoveryState =
        copy(completed = Instant.now())

    override fun asProgress(): Operation.Progress = Operation.Progress(
        started = started,
        total = entities.examined.size,
        processed = entities.processed.size,
        failures = entities.failed.size + failures.size,
        completed = completed
    )

    data class Entities(
        val examined: Set<EntityRef>,
        val collected: Map<EntityRef, TargetEntity>,
        val pending: Map<EntityRef, PendingTargetEntity>,
        val processed: Map<EntityRef, ProcessedTargetEntity>,
        val metadataApplied: Set<EntityRef>,
        val failed: Map<EntityRef, String>
    ) {
        companion object {
            fun empty(): Entities = Entities(
                examined = emptySet(),
                collected = emptyMap(),
                pending = emptyMap(),
                processed = emptyMap(),
                metadataApplied = emptySet(),
                failed = emptyMap()
            )
        }
    }

    data class PendingTargetEntity(
        val expectedParts: Int,
        val processedParts: Int
    ) {
        fun inc(): PendingTargetEntity =
            copy(processedParts = processedParts + 1)
    }

    data class ProcessedTargetEntity(
        val expectedParts: Int,
        val processedParts: Int
    )

    companion object {
        fun start(operation: OperationId): RecoveryState = RecoveryState(
            operation = operation,
            started = Instant.now(),
            entities = Entities.empty(),
            failures = emptyList(),
            completed = null
        )

        fun toProto(
            state: RecoveryState
        ): stasis.client_android.lib.model.proto.RecoveryState =
            stasis.client_android.lib.model.proto.RecoveryState(
                started = state.started.toEpochMilli(),
                entities =
                stasis.client_android.lib.model.proto.RecoveryEntities(
                    examined = state.entities.examined.map { it.key }.toList(),
                    collected = state.entities.collected.map { (k, v) ->
                        k.key to toProtoTargetEntity(v)
                    }.toMap(),
                    pending = state.entities.pending.map { (k, v) ->
                        k.key to toProtoPendingTargetEntity(v)
                    }.toMap(),
                    processed = state.entities.processed.map { (k, v) ->
                        k.key to toProtoProcessedTargetEntity(v)
                    }.toMap(),
                    metadataApplied = state.entities.metadataApplied.map { it.key }.toList(),
                    failed = state.entities.failed.map { (k, v) -> k.key to v }.toMap()
                ),
                failures = state.failures,
                completed = state.completed?.toEpochMilli()
            )

        fun fromProto(
            operation: OperationId, state: stasis.client_android.lib.model.proto.RecoveryState
        ): Try<RecoveryState> =
            when (val entities = state.entities) {
                null -> Try.Failure(IllegalArgumentException("Expected entities in recovery state but none were found"))
                else -> Try {
                    RecoveryState(
                        operation = operation,
                        started = Instant.ofEpochMilli(state.started),
                        entities = Entities(
                            examined = entities.examined.map { EntityRef.default(it) }.toSet(),
                            collected = entities.collected.map { (k, v) ->
                                EntityRef.default(k) to fromProtoTargetEntity(v)
                            }.toMap(),
                            pending = entities.pending.map { (k, v) ->
                                EntityRef.default(k) to fromProtoPendingTargetEntity(v)
                            }.toMap(),
                            processed = entities.processed.map { (k, v) ->
                                EntityRef.default(k) to fromProtoProcessedTargetEntity(v)
                            }.toMap(),
                            metadataApplied = entities.metadataApplied.map { EntityRef.default(it) }.toSet(),
                            failed = entities.failed.map { (k, v) -> EntityRef.default(k) to v }.toMap()
                        ),
                        failures = state.failures,
                        completed = state.completed?.let { Instant.ofEpochMilli(it) }
                    )
                }
            }

        private fun fromProtoTargetEntity(entity: stasis.client_android.lib.model.proto.TargetEntity): TargetEntity =
            TargetEntity(
                ref = EntityRef.default(entity.ref),
                destination = when (val directory = entity.destination?.directory) {
                    null -> TargetEntity.Destination.Default
                    else -> TargetEntity.Destination.Directory(
                        path = Paths.get(directory.path),
                        keepDefaultStructure = directory.keepDefaultStructure,
                        preserveExisting = false
                    )
                },
                existingMetadata = when (val metadata = entity.existingMetadata?.toModel()?.toOption()) {
                    null -> throw IllegalArgumentException("Expected existing metadata in recovery state but none was found")
                    else -> metadata
                },
                currentMetadata = entity.currentMetadata?.let { metadata ->
                    when (val result = metadata.toModel()) {
                        is Try.Success -> result.value
                        is Try.Failure -> throw result.exception
                    }
                }
            )

        private fun toProtoTargetEntity(entity: TargetEntity): stasis.client_android.lib.model.proto.TargetEntity =
            stasis.client_android.lib.model.proto.TargetEntity(
                ref = entity.ref.key,
                destination = when (val destination = entity.destination) {
                    is TargetEntity.Destination.Directory ->
                        stasis.client_android.lib.model.proto.TargetEntityDestination(
                            directory = stasis.client_android.lib.model.proto.TargetEntityDestinationDirectory(
                                path = destination.path.toAbsolutePath().toString(),
                                keepDefaultStructure = destination.keepDefaultStructure
                            )
                        )

                    is TargetEntity.Destination.Default ->
                        stasis.client_android.lib.model.proto.TargetEntityDestination(
                            default = stasis.client_android.lib.model.proto.TargetEntityDestinationDefault()
                        )
                },
                existingMetadata = entity.existingMetadata.toProto(),
                currentMetadata = entity.currentMetadata?.toProto()
            )

        private fun fromProtoPendingTargetEntity(
            entity: stasis.client_android.lib.model.proto.PendingTargetEntity
        ): PendingTargetEntity =
            PendingTargetEntity(
                expectedParts = entity.expectedParts,
                processedParts = entity.processedParts
            )

        private fun toProtoPendingTargetEntity(
            entity: PendingTargetEntity
        ): stasis.client_android.lib.model.proto.PendingTargetEntity =
            stasis.client_android.lib.model.proto.PendingTargetEntity(
                expectedParts = entity.expectedParts,
                processedParts = entity.processedParts
            )

        private fun fromProtoProcessedTargetEntity(
            entity: stasis.client_android.lib.model.proto.ProcessedTargetEntity
        ): ProcessedTargetEntity =
            ProcessedTargetEntity(
                expectedParts = entity.expectedParts,
                processedParts = entity.processedParts
            )

        private fun toProtoProcessedTargetEntity(
            entity: ProcessedTargetEntity
        ): stasis.client_android.lib.model.proto.ProcessedTargetEntity =
            stasis.client_android.lib.model.proto.ProcessedTargetEntity(
                expectedParts = entity.expectedParts,
                processedParts = entity.processedParts
            )
    }
}
